File: steit-csharp/src/Codec/Reader.cs
using System;
using System.Text;

using Steit.State;

namespace Steit.Codec {
    public static class Reader {
        // Wire type occupies three bits.
        public const int WireTypeBits = 3;

        // This mask can be applied to obtain wire type.
        public const UInt32 WireTypeMask = (1U << WireTypeBits) - 1;

        public static bool EndOfStream(this IReader reader) {
            return reader.Remaining() <= 0;
        }

        public static byte[] ReadToEnd(this IReader reader) {
            return reader.Read(reader.Remaining());
        }

        public static ulong ReadUnsignedVarint(this IReader reader) {
            ulong value = 0;

            int offset = 0;
            byte octet;

            while (true) {
                octet = reader.Read();
                value |= (ulong) (octet & 0x7f) << offset;

                if ((octet & 0x80) == 0) {
                    return value;
                }

                offset += 7;
            }
        }

        public static long ReadSignedVarint(this IReader reader) {
            long value = (long) reader.ReadUnsignedVarint();
            return (value >> 1) ^ -(value & 1);
        }

        public static Byte ReadByte(this IReader reader) { return (Byte) reader.ReadUnsignedVarint(); }
        public static UInt16 ReadUInt16(this IReader reader) { return (UInt16) reader.ReadUnsignedVarint(); }
        public static UInt32 ReadUInt32(this IReader reader) { return (UInt32) reader.ReadUnsignedVarint(); }
        public static UInt64 ReadUInt64(this IReader reader) { return reader.ReadUnsignedVarint(); }

        public static SByte ReadSByte(this IReader reader) { return (SByte) reader.ReadSignedVarint(); }
        public static Int16 ReadInt16(this IReader reader) { return (Int16) reader.ReadSignedVarint(); }
        public static Int32 ReadInt32(this IReader reader) { return (Int32) reader.ReadSignedVarint(); }
        public static Int64 ReadInt64(this IReader reader) { return reader.ReadSignedVarint(); }

        public static Boolean ReadBoolean(this IReader reader) {
            bool value = false;
            byte octet;

            while (true) {
                octet = reader.Read();
                value |= (octet & 0x7f) != 0;

                if ((octet & 0x80) == 0) {
                    return value;
                }
            }
        }

        public static String ReadString(this IReader reader) {
            var bytes = reader.GetNested().ReadToEnd();
            return Encoding.UTF8.GetString(bytes);
        }

        public static (UInt32 Tag, WireType WireType) ReadKey(this IReader reader) {
            var key = reader.ReadUInt32();
            var tag = key >> WireTypeBits;
            var wireType = WireTypeHelper.New(key & WireTypeMask);
            return (tag, wireType);
        }

        public static int ReadSize(this IReader reader) {
            return (int) reader.ReadUInt32();
        }

        public static T ReadValue<T>(this IReader reader, Path path, UInt32 tag) {
            return StateFactory.DeserializeNested<T>(reader, path, tag);
        }

        public static void SkipToEnd(this IReader reader) {
            reader.Skip(reader.Remaining());
        }

        public static void SkipField(this IReader reader, WireType wireType) {
            switch (wireType) {
                case WireType.Varint:
                    reader.ReadBoolean();
                    break;

                case WireType.Sized:
                    reader.Skip(reader.ReadSize());
                    break;

                default:
                    throw new NotSupportedException(String.Format("Unsupported wire type: {0}", wireType));
            }
        }

        public static IReader GetNested(this IReader reader) {
            var bytes = reader.Read(reader.ReadSize());
            return new ByteReader(bytes);
        }
    }
}

File: steit-go/pkg/reader/reader.go
package reader

import (
	"fmt"

	"github.com/axieinfinity/steit/steit-go/pkg/codec"
)

const (
	WireTypeBits uint32 = 3
	WireTypeMask uint32 = (uint32(1) << WireTypeBits) - 1
)

type Reader struct {
}

func EndOfStream(reader IReader) bool {
	return reader.Remaining() <= 0
}

func ReadToEnd(reader IReader) []byte {
	return reader.Read(reader.Remaining())
}

func ReadUnsignedVarint(reader IReader) uint64 {
	var value uint64
	offset := 0
	var octet byte

	for {
		octet = reader.ReadUint8()
		value |= uint64((octet & 0x7f)) << offset

		if (octet & 0x80) == 0 {
			return value
		}

		offset += 7
	}
}

func ReadSignedVarint(reader IReader) int64 {
	var value int64
	value = int64(ReadUnsignedVarint(reader))
	return (value >> 1) ^ -(value & 1)
}

func ReadByte(reader IReader) byte {
	return byte(ReadUnsignedVarint(reader))
}

func ReadUInt16(reader IReader) uint16 {
	return uint16(ReadUnsignedVarint(reader))
}

func ReadUInt32(reader IReader) uint32 {
	return uint32(ReadUnsignedVarint(reader))
}

func ReadUInt64(reader IReader) uint64 {
	return ReadUnsignedVarint(reader)
}

func ReadInt16(reader IReader) int16 {
	return int16(ReadSignedVarint(reader))
}

func ReadInt32(reader IReader) int32 {
	return int32(ReadSignedVarint(reader))
}

func ReadInt64(reader IReader) int64 {
	return ReadSignedVarint(reader)
}

func ReadBoolean(reader IReader) bool {
	var value bool
	value = false
	var octet byte

	for {
		octet = reader.ReadUint8()
		value = value || ((octet & 0x7f) != 0)

		if (octet & 0x80) == 0 {
			return value
		}
	}
}

func ReadString(reader IReader) string {
	panic("unimplemented")
}

// Tag(29 bits) | WireType (3 bits)
func ReadKey(reader IReader) (uint32, codec.WireType) {
	key := ReadUInt32(reader)
	tag := key >> WireTypeBits
	wireType := codec.NewWireType(key & WireTypeMask)

	return tag, wireType
}

func ReadSize(reader IReader) int {
	return int(ReadUInt32(reader))
}

func SkipToEnd(reader IReader) {
	reader.Skip(reader.Remaining())
}

func SkipField(reader IReader, wireType codec.WireType) {
	switch wireType {
	case codec.WireTypeVarint:
		ReadBoolean(reader)
	case codec.WireTypeSized:
		SkipToEnd(reader)
	default:
		panic(fmt.Sprintf("Unsupported wire type: %d", wireType))
	}
}

func GetNested(reader IReader) IReader {
	bytes := reader.Read(ReadSize(reader))
	return NewByteReader(bytes)
}
